lambda/dissect_letrec.ml

```diff
@@ -0,0 +1,592 @@
+      | None -> dead_code lam letrec
+    end
+  | Lprim (Pduprecord (kind, size), args, _) -> begin
+      match current_let with
+      | Some cl ->
+          let arg =
+            match args with
+            | [ arg ] -> arg
+            | _ -> Misc.fatal_error "Dissect_letrec.prepare_letrec duprecord"
+          in
+          (match kind with
+           | Types.Record_regular ->
+               build_block cl size (Normal 0) arg letrec
+           | Types.Record_inlined tag ->
+               build_block cl size (Normal tag) arg letrec
+           | Types.Record_extension _ ->
+               build_block cl (size + 1) (Normal 0) arg letrec
+           | Types.Record_unboxed _ ->
+               assert false
+           | Types.Record_float ->
+               build_block cl size Boxed_float arg letrec)
+      | None -> dead_code lam letrec
+    end
+  | Lconst const ->
+      (match current_let with
+       | Some current_let ->
+           { letrec with consts = (current_let.ident, const) :: letrec.consts }
+       | None -> dead_code lam letrec)
+  | Llet (Variable, _, _, _, _) ->
+      (* This is not supposed to appear at this point *)
+      assert false
+  | Llet ((Strict | Alias | StrictOpt) as let_kind, value_kind, id, def, body)
+    ->
+      let letbound = Ident.Set.add id letrec.letbound in
+      let letrec = { letrec with letbound } in
+      let free_vars = Lambda.free_variables def in
+      if Ident.Set.disjoint free_vars recursive_set then
+        (* Non recursive let *)
+        let letrec = prepare_letrec recursive_set current_let body letrec in
+        let pre ~tail : Lambda.lambda =
+          Llet (let_kind, value_kind, id, def, letrec.pre ~tail)
+        in
+        { letrec with pre }
+      else
+      let recursive_set = Ident.Set.add id recursive_set in
+      let letrec = prepare_letrec recursive_set current_let body letrec in
+      let let_def = { let_kind; value_kind; ident = id } in
+      prepare_letrec recursive_set (Some let_def) def letrec
+  | Lsequence (lam1, lam2) ->
+      let letrec = prepare_letrec recursive_set current_let lam2 letrec in
+      prepare_letrec recursive_set None lam1 letrec
+  | Levent (body, event) ->
+      let letrec = prepare_letrec recursive_set current_let body letrec in
+      { letrec with effects = Levent (letrec.effects, event) }
+  | Lletrec (bindings, body) ->
+      (* Inner letrecs need some special care: We split between _outer_ bindings
+         that are recursive with the current [recursive_set], and therefore need
+         to be merged into it, and _inner_ bindings that are not (but can still
+         be recursive between themselves). This corresponds to the [Recursive]
+         case in [Rec_check.value_bindings].
+
+         One solution would be to handle now the outer bindings, and lift the
+         inner ones into [letrec.pre], to be handled in a second pass. That
+         would change the evaluation order, though, so we instead descend
```

Comment:
It is not clear to me why this changes the evaluation order. Generally the rest of the code assumes that any code at the root of the current `lam` argument can be safely (evaluation-order-preservingly) moved into `pre`, and here it seems to not be an issue.

Maybe the problem comes from splitting these inner-bindings in two groups (one for `pre` and one to add to the current workset), which imposes an evaluation order between the two groups that is not the source order?